src/workerd/api/basics.c++

```diff
@@ -521,13 +522,93 @@ bool EventTarget::dispatchEvent(jsg::Lock& js, jsg::Ref<Event> event) {
   return dispatchEventImpl(js, kj::mv(event));
 }
 
+// A wrapper for the AbortTrigger jsrpc client, that automatically sends a release() message once
+// the client is destroyed, informing the server that an abort will not be triggered in the future.
+class AbortTriggerRpcClient final {
+ public:
+  AbortTriggerRpcClient(rpc::AbortTrigger::Client&& client): client(kj::mv(client)) {}
+
+  kj::Promise<void> abort(kj::ArrayPtr<kj::byte> reason) {
+    auto req = client.abortRequest();
+    auto field = req.initReason();
+    field.setV8Serialized(reason);
+    return req.send().ignoreResult();
+  }
+
+  ~AbortTriggerRpcClient() {
+    if (skipReleaseForTest) {
+      return;
+    }
+
+    auto req = client.releaseRequest();
+    // We call detach() on the resulting promise so that we can perform RPC in a destructor
+    req.send().ignoreResult().detach([](kj::Exception exc) {
+      LOG_PERIODICALLY(DBG, "Unable to send AbortTrigger release message", exc);
+    });
+  }
+
+  bool skipReleaseForTest = false;
+
+ private:
+  rpc::AbortTrigger::Client client;
+};
+
+namespace {
+// The jsrpc handler that receives aborts from the remote and triggers them locally
+class AbortTriggerRpcServer final: public rpc::AbortTrigger::Server {
+ public:
+  AbortTriggerRpcServer(kj::Own<kj::PromiseFulfiller<void>> fulfiller,
+      kj::Own<AbortSignal::PendingReason>&& pendingReason)
+      : fulfiller(kj::mv(fulfiller)),
+        pendingReason(kj::mv(pendingReason)) {}
+
+  kj::Promise<void> abort(AbortContext abortCtx) override {
+    auto params = abortCtx.getParams();
+    auto reason = params.getReason().getV8Serialized();
+
+    pendingReason->init<kj::Array<kj::byte>>(kj::heapArray(reason.asBytes()));
+    fulfiller->fulfill();
+
+    abortCtx.initResults();
+    return kj::READY_NOW;
+  }
+
+  kj::Promise<void> release(ReleaseContext releaseCtx) override {
+    released = true;
+    releaseCtx.initResults();
+    return kj::READY_NOW;
+  }
+
+  ~AbortTriggerRpcServer() {
+    if (*pendingReason != nullptr) {
+      // Already triggered
+      return;
+    }
+
+    if (!released) {
+      pendingReason->init<kj::Exception>(kj::Exception(kj::Exception::Type::FAILED, __FILE__,
+          __LINE__, kj::str(JSG_ERROR_DOMAbortError ": Dropped capability, assuming abort")));
+    }
+
+    // Always fulfill the promise in case the AbortSignal was waiting
+    fulfiller->fulfill();
+  }
+
+ private:
+  kj::Own<kj::PromiseFulfiller<void>> fulfiller;
+  kj::Own<AbortSignal::PendingReason> pendingReason;
```

Comment:
Can these be combined into a single `kj::Own<kj::PromiseFulfiller<kj::Own<AbortSignal::PendingReason>>>`? Then you can attach a continuation on the `signal->maybeRpcAbortPromise` that triggers the abort signal.

Reply:
Is there a way to do a non-blocking check for the reason being set, if someone calls AbortSignal.reason or AbortSignal.throeUfAborted() etc?

Reply:
You can't synchronously check a promise, no. I agree these probably have to stay separate like this. Maybe add a comment explaining the motivation, though.